### PR TITLE
Reading first connected device from list

### PR DESCRIPTION
**Pull request by shabbir81**
- Filter list of all devices having Mooltipass Mac-Address
- Filter all devices having Connected State
- Returning first connected device Now

### CODE REVIEW

**Comment by mathfactory** [comment]
Thanks for the PR! Some remarks/questions:
1. What problem does this PR solve? Is it addressing #21, or something else?
2. it seems like "isConnected" is private API in the use of reflection with device.javaClass.getMethod("isConnected") and I am worried that this solution may e.g. not work on all devices and/or android versions.
3. Is evaluating the connection state the right way to solve #21 anyway? I was thinking more of a way to let the user decide which device to connect to (e.g. serial number and/or mac address, or device name(?)) instead of using the isConnected information (which might not represent what the user wants).

**Comment by mathfactory** [comment]
On a second thought I really oppose using private API of the BluetoothDevice class as there are certainly other ways. It seems something along the lines of

`bluetoothManager.getConnectionState(device, BluetoothProfile.GATT) == BluetoothProfile.STATE_CONNECTED`

could do the trick as well? Alternatively, registering a BroadcastReceiver for this information may also be possible. Would you mind checking other options for this? Thanks!

**Comment by shabbir81** [comment]
Solution not working 

- Broadcast Receiver and not found any method that can tell about the connected state of BT device

**Comment by shabbir81** [comment]
Kindly review from last commit e660d29f8c8f508d87639a0b86e7f196aeb044b7

**Comment by limpkin** [comment]
@shabbir81 could you describe the logic of your last commit? does your logic still works if multiple devices are connected **before** mooltifill is launched?
@mathfactory could you give it a look?

**Comment by shabbir81** [comment]
> could you describe the logic of your last commit? does your logic still works if multiple devices are connected **before** mooltifill is launched?

It will capture any connected device having our required mac address series. If we have multiple devices connected, again and again, it will save the last connected device's mac address and when the user tries to test a device connection, it will fetch all paired mooltipass devices and try to connect with a matched device mac.

When mooltifill is launched we don't have any saved last connected device, now it will save any mooltipass connected device in shared preference and use on try a connection button.

**Comment by limpkin** [comment]
pinging @mathfactory for review :) thanks!

File: app/src/main/java/de/mathfactory/mooltifill/MooltipassDevice.kt
Comment (mathfactory) [inline app/src/main/java/de/mathfactory/mooltifill/MooltipassDevice.kt]: You are using Context.registerReceiver() here but I do not see a corresponding Context.unregisterReceiver(), which will leak memory... and the receiver will be registered multiple times on each invocation of pairedDevice(), I think? Also, regarding the registering the receiver through the pairedDevice() function: Did you verify this solution works when e.g. not using the "Mooltipass Awareness" option? I think pairedDevice() in this case will get called not until the instance when device access is made, which might be too late to get the ACTION_ACL_CONNTECTED event.

File: app/src/main/java/de/mathfactory/mooltifill/MooltipassDevice.kt
Comment (mathfactory) [inline app/src/main/java/de/mathfactory/mooltifill/MooltipassDevice.kt]: I would prefer to use "?." instead of "!!.". I don't know if there are cases in which address might be null. If we cannot be sure, it would be better to check all accesses of "BluetootDevice.address" for null with the "?" operator.

File: app/src/main/java/de/mathfactory/mooltifill/MooltipassDevice.kt
Comment (mathfactory) [inline app/src/main/java/de/mathfactory/mooltifill/MooltipassDevice.kt]: Why is it necessary to store the DEVICE_NAME?

File: app/src/main/java/de/mathfactory/mooltifill/MooltipassDevice.kt
Comment (mathfactory) [inline app/src/main/java/de/mathfactory/mooltifill/MooltipassDevice.kt]: Please do not store the context when it is not necessary! It should preferably be passed and I think you could pass it easily to filter() from the pairedDevice(context: Context) function. Please note also my comment about storing the context already in #24.

**Review by mathfactory — CHANGES_REQUESTED** [review]

**Comment by mathfactory** [comment]
Note that in the meantime I have had a look on the issue also, and IMHO giving the user a feedback / choice about the available / connected devices would be a preferable design decision (but I'm open to arguments: the solution here, if fixed, is also much better than before!). I've already implemented a device choice in the settings screen on my machine and would like to start a PR in the next days for discussion.

**Comment by limpkin** [comment]
thanks a lot for your review @mathfactory !
from the few contacts with customers I had on this topic, any request for action from the application is always perceived as intrusive. I'd therefore prefer if by default mooltifill connects to whatever device is connected at this moment. However, we could add an option that would allow the user to select the behavior he'd like :)

File: app/src/main/java/de/mathfactory/mooltifill/MooltipassDevice.kt
Comment (limpkin) [reply app/src/main/java/de/mathfactory/mooltifill/MooltipassDevice.kt]: guessing if we want to present the list at some point, given our latest firmware allows device name customization.

**Comment by limpkin** [comment]
thanks @shabbir81 !
@mathfactory could you give it a final look? :)

**Comment by mathfactory** [comment]
Hi,

I've created a PR #33 with the changes I mentioned before.

> from the few contacts with customers I had on this topic, any request for action from the application is always perceived as intrusive. I'd therefore prefer if by default mooltifill connects to whatever device is connected at this moment. However, we could add an option that would allow the user to select the behavior he'd like :)

In case the user only owns one device, I agree there should not be any "request for action". But when a user owns two (or more) devices, IMO giving the user a choice which device he wants to use is a plus: In the case the user is actively using multiple devices, there is no "magic" way to find out which device they want to use except asking them.

Note that in the PR #33 I decided to implement trying to connect to all bonded devices and picking the first device that responds (if the user didn't choose a specific device), which is an alternative to find out which of the devices is accessible without user intervention.

**Comment by limpkin** [comment]
I feel quite in a tough spot here.
@mathfactory Could you expand on the advantages of your solution to connect to the first device that responds, compared to the observer solution? Is one solution faster than the other? If so, how fast?

File: app/src/main/java/de/mathfactory/mooltifill/SettingsActivity.kt
Comment (mathfactory) [inline app/src/main/java/de/mathfactory/mooltifill/SettingsActivity.kt]: Scheduling the observer in SettingsActivity.onCreate() will only get triggered when the user opens the settings app, which might - after reboot - not be the case at all. Also, onDestroy() might be called pretty soon on low-profile devices, unregistering the receiver. Suggestion: Register the receiver in the android manifest with an intent-filter with an `<action android:name="android.bluetooth.device.action.ACL_CONNECTED" />` instead of in code.

File: app/src/main/java/de/mathfactory/mooltifill/SettingsActivity.kt
Comment (mathfactory) [inline app/src/main/java/de/mathfactory/mooltifill/SettingsActivity.kt]: Suggestion when using the previously mentioned method of registering: Implement the BroadcastReceiver in a new file like the already existing BootReceiver.kt

File: app/src/main/java/de/mathfactory/mooltifill/SettingsActivity.kt
Comment (mathfactory) [inline app/src/main/java/de/mathfactory/mooltifill/SettingsActivity.kt]: Please remove this line (commented code).

File: app/src/main/java/de/mathfactory/mooltifill/MooltipassDevice.kt
Comment (mathfactory) [inline app/src/main/java/de/mathfactory/mooltifill/MooltipassDevice.kt]: Example for when this will fail: The user has already bonded and connected a device, then installs the app -> the connection broadcast event will not be seen, because the app hasn't been installed yet, so isConnected will return false for all `bondedDevices`, leading to no connection being made. If there is has not been a device observed yet, I suggest to fallback to the previous method, i.e.: `return bluetoothManager.adapter.bondedDevices.lastOrNull(::filter)`

**Review by mathfactory — APPROVED** [review]
There still are some issues I noted. For a pragmatic solution and to speed up things I would propose either:
1. If I am allowed I could directly add commits to this PR (not sure if I have permission, there also is an option "Allow edits by maintainers" which probably needs to be checked)
2. Integrate the PR in the current form and I'll start a follow-up PR to address the issues

Of course the creator could try to fix the remaining issues, but I fear this will delay progress unnecessarily.

Further suggestion would be to adapt PR #33 after merging PR #32, if an info screen of available devices would be desired, or to abandon it otherwise.

Approving this PR, with the previous remarks in mind (if you merge, please wait for my follow-up PR before publishing)

**Comment by mathfactory** [comment]
> I feel quite in a tough spot here. @mathfactory Could you expand on the advantages of your solution to connect to the first device that responds, compared to the observer solution? Is one solution faster than the other? If so, how fast?

The difference here is not so much in speed, but more in the method itself.
- Possible drawback of the observer solution is, that we might fail to observe all events in all circumstances, for example when a fresh install happened (which is a critical operation for user experience).
- Possible drawback of connection to all bonded devices simultaneously is maybe higher battery drain for the operation(?), and more code-base changes, which needs more testing.

As indicated in the review notes, it's okay for me to use the observer method and abandon or change my PR accordingly.

File: app/src/main/java/de/mathfactory/mooltifill/SettingsActivity.kt
Comment (shabbir81) [reply app/src/main/java/de/mathfactory/mooltifill/SettingsActivity.kt]: Done in the local commit.

File: app/src/main/java/de/mathfactory/mooltifill/SettingsActivity.kt
Comment (shabbir81) [reply app/src/main/java/de/mathfactory/mooltifill/SettingsActivity.kt]: Done in the local commit.

File: app/src/main/java/de/mathfactory/mooltifill/SettingsActivity.kt
Comment (shabbir81) [reply app/src/main/java/de/mathfactory/mooltifill/SettingsActivity.kt]: Done in the local commit.

File: app/src/main/java/de/mathfactory/mooltifill/MooltipassDevice.kt
Comment (shabbir81) [reply app/src/main/java/de/mathfactory/mooltifill/MooltipassDevice.kt]: In this case, if we don't have any last connected device yet on the device, we can just check the following conditions

- if the device is bonded and the device has the same mac address, in this case, we don't check the last connected state right?

Comment (mathfactory) [reply app/src/main/java/de/mathfactory/mooltifill/MooltipassDevice.kt]: "[...] the device has the same mac address, [...]" - "same" compared to what? In the case we don't have any last connected device address stored, we don't have anything to compare to, do we?

**Comment by limpkin** [comment]
oh, if the connection to all devices is simultaneous in your PR, then maybe we can migrate to it as higher battery drain is not so much of an inconvenience. give me the green light to merge when you successfully tested it? :)

File: app/src/main/java/de/mathfactory/mooltifill/MooltipassDevice.kt
Comment (shabbir81) [reply app/src/main/java/de/mathfactory/mooltifill/MooltipassDevice.kt]: Yes, I have already noticed it and covered it. You will get in the next commit.

**Comment by mathfactory** [comment]
> oh, if the connection to all devices is simultaneous in your PR, then maybe we can migrate to it as higher battery drain is not so much of an inconvenience. give me the green light to merge when you successfully tested it? :)

So, we're talking about #33? I'll have a look then.

**Comment by limpkin** [comment]
Indeed :)

**Pull request closed without merging**
